### PR TITLE
test: add regression tests for RuntimeWarning in H/A mode profiles

### LINKED ISSUE #90: RuntimeWarning: invalid value encountered in power
this line appears to be giving a run time warning

 RuntimeWarning: invalid value encountered in power

    * (1 - (r / pedestal_radius) ** ion_temperature_beta)

I it gets run in 'H' mode

### LINKED ISSUE #115: RuntimeWarning: invalid value encountered in power
I have been running into this issue:

```
RuntimeWarning: invalid value encountered in power
  * (1 - (np.abs(r / pedestal_radius)) ** ion_temperature_beta)
```

When running the tokamak_source_example.py or other custom parameters:

```
my_source = tokamak_source(
    elongation=1.739,
    ion_density_centre=6.8e19,
    ion_density_pedestal=5.78e19,
    ion_density_peaking_factor=1,
    ion_density_separatrix=1.02e19,
    ion_temperature_centre=23.7e3,
    ion_temperature_pedestal=5.5e3,
    ion_temperature_separatrix=0.1e3,
    ion_temperature_peaking_factor=8.06,
    ion_temperature_beta=6,
    major_radius=840.67,
    minor_radius=300.2,
    pedestal_radius=0.94 * 300.2,
    mode="H",
    shafranov_factor=0.44789,
    triangularity=0.333,
    fuel={"D": 0.5, "T": 0.5},
)
```

### PR DESCRIPTION
**Pull request by steps-re**
## Summary

Adds regression tests that pin the fix for [#90] and [#115] — two open
issues reporting `RuntimeWarning: invalid value encountered in power` when
calling `tokamak_ion_temperature` (and implicitly `tokamak_source`) in
H-mode with realistic parameters such as `ion_temperature_beta=6`.

**Root cause.** `np.where` evaluates *both* branches for all array
elements before selecting. In the H/A core branch:
```python
(1 - (r / pedestal_radius) ** ion_temperature_beta) ** ion_temperature_peaking_factor
```
elements where `r > pedestal_radius` compute a negative base raised to a
non-integer power, which produces `NaN` and a `RuntimeWarning`. The
existing fix clips with `r_core = np.minimum(r, pedestal_radius)` before
the exponentiation, but no test was verifying that the warning stays gone.

## What this PR adds

**RuntimeWarning regression tests** (close #90 and #115):
- `test_ion_density_h_a_mode_no_warning` — verifies no `RuntimeWarning`
  when calling `tokamak_ion_density` in H and A mode over the full
  `[0, minor_radius]` range.
- `test_ion_temperature_h_a_mode_no_warning` — same for
  `tokamak_ion_temperature`.
- `test_tokamak_source_h_a_mode_no_warning` — same for the full
  `tokamak_source` call path.

**Physical boundary-condition tests**:
- `test_ion_density_all_modes_finite` — density is finite and
  non-negative across all three modes.
- `test_ion_density_boundary_conditions` — density at `r=0` equals
  `ion_density_centre`; at `r=minor_radius` equals `ion_density_separatrix`
  (H/A) or 0 (L mode).
- `test_ion_temperature_all_modes_finite` — temperature is finite and
  non-negative across all modes.
- `test_ion_temperature_h_a_boundary_conditions` — temperature at `r=0`
  equals `ion_temperature_centre` (converted to eV); at `r=minor_radius`
  equals `ion_temperature_separatrix` (in eV).

These boundary-condition tests also serve as sanity checks that the H/A
mode formulae from Fausser et al. (2012) are implemented correctly end-to-end.

Closes #90
Closes #115

### CODE REVIEW

File: tests/test_tokamak_source.py [hunk 1]
```diff
@@ -529,3 +529,193 @@ def test_strengths_are_volume_weighted(tokamak_args_dict):
...
     assert weighted_mean_R > tokamak_args_dict["major_radius"]
+
+
+# --- Regression tests for RuntimeWarning in H/A mode profiles (issues #90, #115) ---
+#
+# np.where evaluates *both* branches for all elements before selecting. In H/A
+# mode the core branch computes (1 - (r/pedestal_radius)**beta)**alpha; for
+# r > pedestal_radius this raises a negative base to a non-integer power,
+# producing NaN and a RuntimeWarning. The fix clips r in the core branch with
+# r_core = np.minimum(r, pedestal_radius). The tests below pin that fix.
```
Comment (approx-infinity) [inline tests/test_tokamak_source.py]: I think these comments should be removed.

File: tests/test_tokamak_source.py [hunk 2]
```diff
@@ -529,3 +529,193 @@ def test_strengths_are_volume_weighted(tokamak_args_dict):
...
+
+
+@pytest.mark.parametrize("mode", ["H", "A"])
+def test_ion_density_h_a_mode_no_warning(tokamak_args_dict, mode, recwarn):
+    """tokamak_ion_density in H/A mode must not raise a RuntimeWarning.
+
+    Regression for issue #90 and #115: np.where evaluates both branches for
+    all r, so without clipping r the core-branch exponentiation fires for
+    r > pedestal_radius and produces NaN + RuntimeWarning.
```
Comment (approx-infinity) [inline tests/test_tokamak_source.py]:
```
Regression for issue #90 and #115: np.where evaluates both branches for
    all r, so without clipping r the core-branch exponentiation fires for
    r > pedestal_radius and produces NaN + RuntimeWarning.
```
These lines need to be modified. We don't need to mention the issue numbers.

File: tests/test_tokamak_source.py [hunk 3]
```diff
@@ -529,3 +529,193 @@ def test_strengths_are_volume_weighted(tokamak_args_dict):
...
+    )
+    runtime_warnings = [w for w in recwarn.list if issubclass(w.category, RuntimeWarning)]
+    assert runtime_warnings == [], f"Unexpected RuntimeWarnings: {runtime_warnings}"
+
+
+@pytest.mark.parametrize("mode", ["H", "A"])
+def test_ion_temperature_h_a_mode_no_warning(tokamak_args_dict, mode, recwarn):
+    """tokamak_ion_temperature in H/A mode must not raise a RuntimeWarning.
+
+    Regression for issue #90 and #115.
```
Comment (approx-infinity) [inline tests/test_tokamak_source.py]: Same thing here.

File: tests/test_tokamak_source.py [hunk 4]
```diff
@@ -529,3 +529,193 @@ def test_strengths_are_volume_weighted(tokamak_args_dict):
...
+    runtime_warnings = [w for w in recwarn.list if issubclass(w.category, RuntimeWarning)]
+    assert runtime_warnings == [], f"Unexpected RuntimeWarnings: {runtime_warnings}"
+
+
+@pytest.mark.parametrize("mode", ["H", "A"])
+def test_tokamak_source_h_a_mode_no_warning(tokamak_args_dict, mode, recwarn):
+    """tokamak_source in H/A mode must not raise a RuntimeWarning.
+
+    Regression for issue #90 and #115: exercises the full source-creation path
+    which calls tokamak_ion_density and tokamak_ion_temperature internally.
```
Comment (approx-infinity) [inline tests/test_tokamak_source.py]: needs the same modification

File: tests/test_tokamak_source.py [hunk 5]
```diff
@@ -529,3 +529,193 @@ def test_strengths_are_volume_weighted(tokamak_args_dict):
...
+    """
+    tokamak_args_dict["mode"] = mode
+    tokamak_source(**tokamak_args_dict)
+    runtime_warnings = [w for w in recwarn.list if issubclass(w.category, RuntimeWarning)]
+    assert runtime_warnings == [], f"Unexpected RuntimeWarnings: {runtime_warnings}"
+
+
+# --- Physical boundary-condition tests for ion density ---
```
Comment (approx-infinity) [inline tests/test_tokamak_source.py:603]: I think we don't need this

File: tests/test_tokamak_source.py [hunk 6]
```diff
@@ -529,3 +529,193 @@ def test_strengths_are_volume_weighted(tokamak_args_dict):
...
+    np.testing.assert_allclose(density_at_centre[0], ion_density_centre, rtol=1e-6)
+    if mode in ["H", "A"]:
+        np.testing.assert_allclose(
+            density_at_separatrix[0], ion_density_separatrix, rtol=1e-6
+        )
+    else:  # L mode: density drops to 0 at the separatrix
+        np.testing.assert_allclose(density_at_separatrix[0], 0.0, atol=1e-6)
+
+
+# --- Physical boundary-condition tests for ion temperature ---
```
Comment (approx-infinity) [inline tests/test_tokamak_source.py:665]: This can also be removed

**Comment by steps-re** [comment]
Thanks for the review @approx-infinity. All of these are addressed in ecb0536: removed the explanatory block comment, dropped the issue-number references, removed the `# ---` section-divider comments, and trimmed the docstrings to one line. The two later comments landed against the pre-cleanup snapshot, so the current diff should already be clean. Let me know if I missed anything.

**Comment by approx-infinity** [comment]
This looks good to me. However, I don't have permission to merge PRs in this repository, so you'll need to wait until one of the maintainers reviews and merges it.

**Comment by shimwell** [comment]
thanks for the PR and review both. much appreciated. CI is running
